lib/colors.nix

```diff
@@ -69,57 +80,204 @@ let
-      splitRGB' = f: hex: prefix:
-        let rgb = splitRGB hex;
-        in {
-          "${prefix}-r" = f rgb.r;
-          "${prefix}-g" = f rgb.g;
-          "${prefix}-b" = f rgb.b;
-        };
-      addRGB = f: prefix:
-        builtins.foldl' (x: y: x // y) { }
-        (builtins.map (baseXX: splitRGB' f base.${baseXX} "${baseXX}-${prefix}")
-          (builtins.attrNames base));
+      inherit (builtins) listToAttrs attrValues;
+      uncurry = lib.foldl lib.id;
+      mapAttrValues = compose [
+        lib.const
+        lib.mapAttrs
+      ];
+      _color =
+        hex:
+        lib.fix (self: {
+          hex = {
+            __toString = self: hex;
+            inherit (splitRGB self.hex) r g b;
+            bgr = with self.hex; b + g + r;
+            withHashtag = "#${self.hex}";
+          };
+          __toString = self: self.hex;
+          rgb = mapAttrValues toIntBase16 (splitRGB self.hex);
+          dec = mapAttrValues (x: x / 255.0) self.rgb;
+        });
 
-      # populate the fields:
-      base-hex-rgb = addRGB (x: x) "hex";
-      base-rgb-rgb = addRGB (x: builtins.toString (primaryHex2Dec x)) "rgb";
-      base-dec-rgb = addRGB (x: builtins.toString (primaryHex2Dec x / 256.0)) "dec";
+      ansi-list =
+        colors: with colors; [
+          black
+          red
+          green
+          yellow
+          blue
+          magenta
+          cyan
+          white
+        ];
+      ansi =
+        palette:
+        with mkBase24 palette;
+        lib.fix (self: {
+          dark = {
+            black = base01;
+            red = base08;
+            green = base0B;
+            yellow = base09;
+            blue = base0D;
+            magenta = base0E;
+            cyan = base0C;
+            white = base06;
+            # suitable for programs.wezterm.colorSchemes.<name>.ansi and similar
+            toList = ansi-list self.dark;
+          };
+          bright = {
+            black = base02;
+            red = base12;
+            green = base14;
+            yellow = base13;
+            blue = base16;
+            magenta = base17;
+            cyan = base15;
+            white = base07;
+            toList = ansi-list self.bright;
+          };
+          # suitable for console.colors and similar
+          toList = self.dark.toList ++ self.bright.toList;
+        });
 
-      base-hex = lib.mapAttrs' (k: v: lib.nameValuePair "${k}-hex" v) base;
-      base-short = lib.mapAttrs' (k: v: lib.nameValuePair "${k}" v) base;
+      mnemonic =
+        palette:
+        let
+          ansi' = ansi palette;
+        in
+        {
+          inherit (ansi'.dark)
+            red
+            green
+            cyan
+            blue
+            magenta
+            ;
+          orange = ansi'.dark.yellow;
+          yellow = palette.base0A;
+          brown = palette.base0F;
+        }
+        // builtins.listToAttrs (
+          lib.zipListsWith lib.nameValuePair
+            (map (x: "bright-${x}") [
```

Comment:
just prepend bright- to each color

Reply:
This is exactly what happens, just without boilerplate

Reply:
Sadly, the code does not have any tests so more boilerplate is better than more logic